kvbc/include/pruning_reserved_pages_client.hpp

```diff
@@ -0,0 +1,103 @@
+// Concord
+//
+// Copyright (c) 2021 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").  You may not use this product except in
+// compliance with the Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright notices and license terms. Your use of
+// these subcomponents is subject to the terms and conditions of the sub-component's license, as noted in the LICENSE
+// file.
+
+#pragma once
+
+#include "ReservedPagesClient.hpp"
+
+#include "kv_types.hpp"
+
+#include "pruning_msgs.cmf.hpp"
+
+#include <cstdint>
+#include <chrono>
+
+namespace concord::kvbc::pruning {
+
+static constexpr auto kPruningResPageCount = 2;
+
+// Creates a pruning agreement.
+// `tick_period` must be at least 1 second.
+// `batch_blocks_num` must be non-zero.
+// `last_agreed_prunable_block_id` must be at least INITIAL_GENESIS_BLOCK_ID.
+Agreement createAgreement(const std::chrono::seconds& tick_period,
+                          std::uint64_t batch_blocks_num,
+                          BlockId last_agreed_prunable_block_id);
+
+// Provides persistence of pruning agreement and progress information in reserved pages.
+//
+// Pseudo code for how it can be used:
+//
+// * On PruneRequest:
+//   client.saveAgreement(agreement);
+//   if (genesis <= agreement.last_agreed_prunable_block_id) ticks_gen.start();
+//
+// * On Tick:
+//   auto agreement = client.lastAgreement();
+//   if (agreement) {
+//     auto until = std::min(genesis + agreement->batch_blocks_num, agreement->last_agreed_prunable_block_id + 1);
+//     if (until >= genesis) {
+//       client.saveLatestBatch(until - 1);
+//       blockchain.deleteBlocksUntil(until);
+//     }
+//     if (genesis >= agreement->last_agreed_prunable_block_id) ticks_gen.stop();
+//   }
+//
+// * On Startup and State Transfer:
+//   auto agreement = client.lastAgreement();
+//   const auto to = client.latestBatchBlockIdTo();
+//   if (to && *to + 1 > genesis) blockchain.deleteUntil(*to + 1);
+//   if (genesis < agreement->last_agreed_prunable_block_id) ticks_gen.start();
+//   else ticks_gen.stop();
+class ReservedPagesClient : private bftEngine::ResPagesClient<ReservedPagesClient, kPruningResPageCount> {
+ public:
+  using bftEngine::ResPagesClient<ReservedPagesClient, kPruningResPageCount>::getNumResPages;
```

Comment:
I didn't find its usage

Reply:
It is used by `ReservedPagesMock`.

Reply:
I mean this particular line in pruning ReservedPagesClient is not in use. Mock is a non-related class

Reply:
Ah, yes. What I meant is that the inheritance is `private` as I don't think the pruning client needs any of the ResPagesClient methods. That's why I "lift" this method here so that `ReservedPagesMock` can use it. Does that make sense?

Reply:
Or I can just use composition and implement that method. No actual reason for using `private` inheritance here, as far as I remember.

Reply:
yeah, I never found a private inheritance to be really useful